### ISSUE TITLE
[ Question ] Isn't the mid-training evaluation of the SWA performance corrupting the batchnorm running averages ?

### ISSUE THREAD

**Issue by alexandrebone** [timgaripov/swa#15]
Hello, 

Thank you for your code. 

I believe some mechanism is needed to reset the batchnorm running averages to their original values after a mid-training evaluation of the SWA performance (i.e. use of the `bn_update` method). 
Otherwise, evaluating the SWA performance is tampering with the training procedure. 

Maybe a buffer could be created when calling `bn_update`, and then used when the user calls a new method in the flavour of `bn_reset` before resuming to the training ? 

Best, 
Alex

**Comment by timgaripov** [comment]
Hi Alex,

Thank you for your question.

The mid-training evaluation of the SWA performance does not really have any effect on the training. We only apply `bn_update` to `swa_model` which is separated from `model` trained by SGD. Essentially, parameters of `swa_model` implement the buffers where we store the averaged weights of SGD iterates.

Let me know if that answers your question.

Best regards,
Timur

**Comment by timgaripov** [comment]
Hi Alex,

Thank you for your question.

The mid-training evaluation of the SWA performance does not really have any effect on the training. We only apply `bn_update` to `swa_model` which is separated from `model` trained by SGD. Essentially, parameters of `swa_model` implement the buffers where we store the averaged weights of SGD iterates.

Let me know if that answers your question.

Best regards,
Timur

**Comment by alexandrebone** [comment]
Hello Timur, 

Sorry I had not seen your response. Thank you very much for it. 

I think the buffers are only implemented for the parameters _directly_ optimized by the chosen optimizer (i.e. optimized with a gradient-based approach). This would exclude the running averages of the batch norm layers. 
Therefore, I believe after calling the `bn_update` method in the middle of the training, the running averages of batch norm layers are not reset to their original values and are therefore left "corrupted". 

Are some buffers really dedicated to the batch norm running averages ? 

Best, 
Alex 

```
 def swap_swa_sgd(self):
    r"""Swaps the values of the optimized variables and swa buffers.

    It's meant to be called in the end of training to use the collected
    swa running averages. It can also be used to evaluate the running
    averages during training; to continue training `swap_swa_sgd`
    should be called again.
    """
    for group in self.param_groups:  # in init: "self.param_groups = self.optimizer.param_groups"
        for p in group['params']:
            param_state = self.state[p]
            if 'swa_buffer' not in param_state:
                # If swa wasn't applied we don't swap params
                warnings.warn(
                    # "SWA wasn't applied to param {}; skipping it".format(p))
                    "SWA wasn't applied; skipping")
                continue
            buf = param_state['swa_buffer']
            tmp = torch.empty_like(p.data)
            tmp.copy_(p.data)
            p.data.copy_(buf)
            buf.copy_(tmp)
```

**Comment by izmailovpavel** [comment]
Hey @alexandrebone, we realized you are asking about the [`torchcontrib` SWA implementation](https://github.com/pytorch/contrib) and not the implementation in this repo. In any case, in the `torchcontrib` implementation we assume that the non-SWA model is not going to be used after training. The batchnorm statistics do not affect the behavior of the model in train mode, as in train mode the batchnorm layers use batch statistics. So, the behaviour of the non-SWA model is unchanged by calling `swap_swa_sgd` and `bn_udate`.

**Comment by alexandrebone** [comment]
Thank you very much for your answer ! Very clear. 

Best, 
Alex
